test/types/peg.test-d.ts

```diff
@@ -0,0 +1,330 @@
+  });
+
+  it("takes a valid tracer", () => {
+    const parser = peggy.generate(src, { trace: true });
+    expectType<peggy.Parser>(parser);
+
+    const res = parser.parse(" /**/ 1\n", {
+      startRule: "top",
+      tracer: {
+        trace(event) {
+          expectType<peggy.ParserTracerEvent>(event);
+          expectType<"rule.enter" | "rule.match" | "rule.fail">(event.type);
+          expectType<string>(event.rule);
+          switch (event.type) {
+            case "rule.enter":
+              expectType<peggy.LocationRange>(event.location);
+              break;
+            case "rule.match":
+              expectType<peggy.LocationRange>(event.location);
+              expectType<any>(event.result);
+              break;
+          }
+        }
+      }
+    })
+  });
+
+  it("takes an output and grammarSource", () => {
+    const p1 = peggy.generate(src, { output: "parser", grammarSource: "src" });
+    expectType<peggy.Parser>(p1);
+
+    const p2 = peggy.generate(src, { output: "source" });
+    expectType<string>(p2);
+  });
+
+  it("creates an AST", () => {
+    const grammar = peggy.parser.parse(src);
+    expectType<peggy.ast.Grammar>(grammar);
+
+    const visited: { [key: string]: number; } = {};
+    function add(typ:string): void {
+      if (!visited[typ]) {
+        visited[typ] = 1;
+      } else {
+        visited[typ]++
+      }
+    }
+
+    const visit = peggy.compiler.visitor.build({
+      grammar(node) {
+        add(node.type);
+        expectType<peggy.ast.Grammar>(node);
+        expectType<"grammar">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<peggy.ast.TopLevelInitializer|undefined>(node.topLevelInitializer);
+        expectType<peggy.ast.Initializer|undefined>(node.initializer);
+        expectType<peggy.ast.Rule[]>(node.rules);
+
+        if (node.topLevelInitializer) {
+          visit(node.topLevelInitializer);
+        }
+        if (node.initializer) {
+          visit(node.initializer);
+        }
+        node.rules.forEach(visit);
+      },
+      top_level_initializer(node) {
+        add(node.type);
+        expectType<peggy.ast.TopLevelInitializer>(node);
+        expectType<"top_level_initializer">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<string>(node.code);
+        expectType<peggy.LocationRange>(node.codeLocation);
+      },
+      initializer(node) {
+        add(node.type);
+        expectType<peggy.ast.Initializer>(node);
+        expectType<"initializer">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<string>(node.code);
+        expectType<peggy.LocationRange>(node.codeLocation);
+      },
+      rule(node) {
+        add(node.type);
+        expectType<peggy.ast.Rule>(node);
+        expectType<"rule">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<peggy.LocationRange>(node.nameLocation);
```

Comment:
You forgot to check a `name`, if that even needed:
```suggestion
        expectType<peggy.LocationRange>(node.location);
        expectType<string>(node.name);
        expectType<peggy.LocationRange>(node.nameLocation);
```